src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -31,6 +31,62 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region InodeTracker
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<(UInt64, UInt64), bool>  _visitations;
+
+        /// <summary>
+        /// Construct a new InodeTracker
+        /// </summary>
+        public InodeTracker()
+        {
+            _visitations = new Dictionary<(UInt64, UInt64), bool>();
+        }
+
+        /// <summary>
+        /// Determine if a path has been visited.
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be checked.</param>
+        /// <returns>True if the path has been visited, false otherwise</returns>
+        public bool Visited(string path)
+        {
+            var inodeData = (0UL, 0UL);
+
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
+            {
+                return _visitations.ContainsKey(inodeData);
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Mark a path as visited.
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be marked as visited.</param>
+        /// <returns>True if the path was successfully added to the visited list, false otherwise</returns>
+        public bool Visit(string path)
```

Comment:
And, does this method really need to return anything? If `GetInodeData` failed, then we just don't skip anything.